spec/models/user_spec.rb

```diff
@@ -22,10 +22,11 @@
         end
 
         context 'email is present' do
-          it 'generates a confirmation token' do
-            expect(subject).to receive(:generate_confirmation_token)
-            subject.save
-          end
+          it 'generates a confirmation token'
+          # do
+          #   expect(subject).to receive(:generate_confirmation_token)
+          #   subject.save
+          # end
```

Comment:
This test started to fail because it passes user without confirmation token to mailer template ( the previous implementation allowed that to happen without error).

I tried to debug with `byebug`. Breakpoint in Mailer `confirmation` method stopped execution. Going up the stack ended in user model, inside `send_confirmation_email`.

Then I set two additional breakpoints in user model, `send_confirmation_email` and `generate_confirmation_token` but they didn't stop the execution.

I completely don't understand how this is possible.

Error
---
```
Failures:

  1) User callbacks before_save .check_email_changed email is present generates a confirmation token
     Failure/Error: <%= link_to confirm_email_url(@user.confirmation_token), confirm_email_url(@user.confirmation_token) %>

     ActionView::Template::Error:
       No route matches {:action=>"confirm_email", :confirmation_token=>nil, :controller=>"dashboards"}, possible unmatched constraints: [:confirmation_token]
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/actionpack-5.1.4/lib/action_dispatch/journey/formatter.rb:55:in `generate'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/actionpack-5.1.4/lib/action_dispatch/routing/route_set.rb:738:in `generate'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/actionpack-5.1.4/lib/action_dispatch/routing/route_set.rb:769:in `generate'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/actionpack-5.1.4/lib/action_dispatch/routing/route_set.rb:816:in `url_for'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/actionpack-5.1.4/lib/action_dispatch/routing/route_set.rb:267:in `call'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/actionpack-5.1.4/lib/action_dispatch/routing/route_set.rb:208:in `call'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/actionpack-5.1.4/lib/action_dispatch/routing/route_set.rb:325:in `block (2 levels) in define_url_helper'
     # ./app/views/confirmation_mailer/confirmation.html.erb:4:in `_app_views_confirmation_mailer_confirmation_html_erb___185991304532696001_70260479705380'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/actionview-5.1.4/lib/action_view/template.rb:157:in `block in render'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/activesupport-5.1.4/lib/active_support/notifications.rb:168:in `instrument'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/actionview-5.1.4/lib/action_view/template.rb:352:in `instrument_render_template'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/actionview-5.1.4/lib/action_view/template.rb:155:in `render'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/actionview-5.1.4/lib/action_view/renderer/template_renderer.rb:52:in `block (2 levels) in render_template'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/actionview-5.1.4/lib/action_view/renderer/abstract_renderer.rb:42:in `block in instrument'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/activesupport-5.1.4/lib/active_support/notifications.rb:166:in `block in instrument'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/activesupport-5.1.4/lib/active_support/notifications/instrumenter.rb:21:in `instrument'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/activesupport-5.1.4/lib/active_support/notifications.rb:166:in `instrument'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/actionview-5.1.4/lib/action_view/renderer/abstract_renderer.rb:41:in `instrument'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/actionview-5.1.4/lib/action_view/renderer/template_renderer.rb:51:in `block in render_template'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/actionview-5.1.4/lib/action_view/renderer/template_renderer.rb:59:in `render_with_layout'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/actionview-5.1.4/lib/action_view/renderer/template_renderer.rb:50:in `render_template'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/actionview-5.1.4/lib/action_view/renderer/template_renderer.rb:14:in `render'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/actionview-5.1.4/lib/action_view/renderer/renderer.rb:42:in `render_template'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/actionview-5.1.4/lib/action_view/renderer/renderer.rb:23:in `render'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/actionview-5.1.4/lib/action_view/rendering.rb:103:in `_render_template'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/actionview-5.1.4/lib/action_view/rendering.rb:83:in `render_to_body'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/actionpack-5.1.4/lib/abstract_controller/rendering.rb:24:in `render'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/actionmailer-5.1.4/lib/action_mailer/base.rb:935:in `block in collect_responses_from_templates'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/actionmailer-5.1.4/lib/action_mailer/base.rb:932:in `each'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/actionmailer-5.1.4/lib/action_mailer/base.rb:932:in `each'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/actionmailer-5.1.4/lib/action_mailer/base.rb:932:in `map'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/actionmailer-5.1.4/lib/action_mailer/base.rb:932:in `collect_responses_from_templates'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/actionmailer-5.1.4/lib/action_mailer/base.rb:917:in `collect_responses'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/actionmailer-5.1.4/lib/action_mailer/base.rb:830:in `mail'
     # ./app/mailers/confirmation_mailer.rb:8:in `confirmation'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/actionpack-5.1.4/lib/abstract_controller/base.rb:186:in `process_action'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/actionpack-5.1.4/lib/abstract_controller/callbacks.rb:20:in `block in process_action'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/activesupport-5.1.4/lib/active_support/callbacks.rb:97:in `run_callbacks'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/actionpack-5.1.4/lib/abstract_controller/callbacks.rb:19:in `process_action'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/actionpack-5.1.4/lib/abstract_controller/base.rb:124:in `process'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/actionmailer-5.1.4/lib/action_mailer/rescuable.rb:23:in `block in process'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/actionmailer-5.1.4/lib/action_mailer/rescuable.rb:15:in `handle_exceptions'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/actionmailer-5.1.4/lib/action_mailer/rescuable.rb:22:in `process'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/actionview-5.1.4/lib/action_view/rendering.rb:30:in `process'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/actionmailer-5.1.4/lib/action_mailer/base.rb:609:in `block in process'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/activesupport-5.1.4/lib/active_support/notifications.rb:166:in `block in instrument'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/activesupport-5.1.4/lib/active_support/notifications/instrumenter.rb:21:in `instrument'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/activesupport-5.1.4/lib/active_support/notifications.rb:166:in `instrument'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/actionmailer-5.1.4/lib/action_mailer/base.rb:608:in `process'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/actionmailer-5.1.4/lib/action_mailer/message_delivery.rb:105:in `block in processed_mailer'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/actionmailer-5.1.4/lib/action_mailer/message_delivery.rb:104:in `tap'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/actionmailer-5.1.4/lib/action_mailer/message_delivery.rb:104:in `processed_mailer'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/actionmailer-5.1.4/lib/action_mailer/message_delivery.rb:95:in `deliver_now'
     # ./app/models/user.rb:259:in `check_email_changed'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/activesupport-5.1.4/lib/active_support/callbacks.rb:413:in `block in make_lambda'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/activesupport-5.1.4/lib/active_support/callbacks.rb:197:in `block (2 levels) in halting'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/activesupport-5.1.4/lib/active_support/callbacks.rb:601:in `block (2 levels) in default_terminator'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/activesupport-5.1.4/lib/active_support/callbacks.rb:600:in `catch'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/activesupport-5.1.4/lib/active_support/callbacks.rb:600:in `block in default_terminator'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/activesupport-5.1.4/lib/active_support/callbacks.rb:198:in `block in halting'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/activesupport-5.1.4/lib/active_support/callbacks.rb:507:in `block in invoke_before'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/activesupport-5.1.4/lib/active_support/callbacks.rb:507:in `each'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/activesupport-5.1.4/lib/active_support/callbacks.rb:507:in `invoke_before'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/activesupport-5.1.4/lib/active_support/callbacks.rb:130:in `run_callbacks'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/activesupport-5.1.4/lib/active_support/callbacks.rb:827:in `_run_save_callbacks'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/activerecord-5.1.4/lib/active_record/callbacks.rb:336:in `create_or_update'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/activerecord-5.1.4/lib/active_record/persistence.rb:129:in `save'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/activerecord-5.1.4/lib/active_record/validations.rb:44:in `save'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/activerecord-5.1.4/lib/active_record/attribute_methods/dirty.rb:35:in `save'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/activerecord-5.1.4/lib/active_record/transactions.rb:308:in `block (2 levels) in save'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/activerecord-5.1.4/lib/active_record/transactions.rb:384:in `block in with_transaction_returning_status'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/activerecord-5.1.4/lib/active_record/connection_adapters/abstract/database_statements.rb:235:in `block in transaction'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/activerecord-5.1.4/lib/active_record/connection_adapters/abstract/transaction.rb:194:in `block in within_new_transaction'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/activerecord-5.1.4/lib/active_record/connection_adapters/abstract/transaction.rb:191:in `within_new_transaction'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/activerecord-5.1.4/lib/active_record/connection_adapters/abstract/database_statements.rb:235:in `transaction'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/activerecord-5.1.4/lib/active_record/transactions.rb:210:in `transaction'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/activerecord-5.1.4/lib/active_record/transactions.rb:381:in `with_transaction_returning_status'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/activerecord-5.1.4/lib/active_record/transactions.rb:308:in `block in save'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/activerecord-5.1.4/lib/active_record/transactions.rb:323:in `rollback_active_record_state!'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/activerecord-5.1.4/lib/active_record/transactions.rb:307:in `save'
     # /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/activerecord-5.1.4/lib/active_record/suppressor.rb:42:in `save'
     # ./spec/models/user_spec.rb:27:in `block (6 levels) in <top (required)>'
     # ------------------
     # --- Caused by: ---
     # ActionController::UrlGenerationError:
     #   No route matches {:action=>"confirm_email", :confirmation_token=>nil, :controller=>"dashboards"}, possible unmatched constraints: [:confirmation_token]
     #   /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/actionpack-5.1.4/lib/action_dispatch/journey/formatter.rb:55:in `generate'
```

Debugging session
---
```
[3, 12] in /Users/lokson/24pullrequests/app/mailers/confirmation_mailer.rb
    3:
    4:   default from: '24 Pull Requests <info@24pullrequests.com>'
    5:
    6:   def confirmation(user)
    7:     byebug
=>  8:     @user = user
    9:     mail :to         => user.email,
   10:          :subject    => '[24 Pull Requests] Email Confirmation',
   11:          'X-SMTPAPI' => '{"category": "Email Confirmation"}'
   12:   end
(byebug) user
#<User id: nil, uid: 1, provider: "github", nickname: "eileen-beahan", email: "email1@factory.com", created_at: nil, updated_at: nil, gravatar_id: "dedrick_wisoky@wisozk.com", token: "97139babbec73b112ae5801ea2ce345d", email_frequency: nil, pull_requests_count: 0, last_sent_at: nil, twitter_token: nil, twitter_secret: nil, twitter_nickname: nil, confirmation_token: nil, confirmed_at: nil, name: nil, blog: nil, location: "Cambodia", thank_you_email_sent: false, lat: 0.40714353e2, lng: -0.74005973e2, ignored_organisations: [], unsubscribe_token: "ec388d37-7b77-4ee0-a308-fa6650b18797">
(byebug) up

[181, 190] in /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/actionpack-5.1.4/lib/abstract_controller/base.rb
   181:       # is the intended way to override action dispatching.
   182:       #
   183:       # Notice that the first argument is the method to be dispatched
   184:       # which is *not* necessarily the same as the action name.
   185:       def process_action(method_name, *args)
=> 186:         send_action(method_name, *args)
   187:       end
   188:
   189:       # Actually call the method associated with the action. Override
   190:       # this method if you wish to change how action methods are called,
(byebug) up

[15, 24] in /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/actionpack-5.1.4/lib/abstract_controller/callbacks.rb
   15:
   16:     # Override AbstractController::Base's process_action to run the
   17:     # process_action callbacks around the normal behavior.
   18:     def process_action(*args)
   19:       run_callbacks(:process_action) do
=> 20:         super
   21:       end
   22:     end
   23:
   24:     module ClassMethods
(byebug) up

[92, 101] in /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/activesupport-5.1.4/lib/active_support/callbacks.rb
    92:     # as possible that we were here.
    93:     def run_callbacks(kind)
    94:       callbacks = __callbacks[kind.to_sym]
    95:
    96:       if callbacks.empty?
=>  97:         yield if block_given?
    98:       else
    99:         env = Filters::Environment.new(self, false, nil)
   100:         next_sequence = callbacks.compile
   101:
(byebug) up

[14, 23] in /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/actionpack-5.1.4/lib/abstract_controller/callbacks.rb
   14:     end
   15:
   16:     # Override AbstractController::Base's process_action to run the
   17:     # process_action callbacks around the normal behavior.
   18:     def process_action(*args)
=> 19:       run_callbacks(:process_action) do
   20:         super
   21:       end
   22:     end
   23:
(byebug) up

[119, 128] in /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/actionpack-5.1.4/lib/abstract_controller/base.rb
   119:         raise ActionNotFound, "The action '#{action}' could not be found for #{self.class.name}"
   120:       end
   121:
   122:       @_response_body = nil
   123:
=> 124:       process_action(action_name, *args)
   125:     end
   126:
   127:     # Delegates to the class' ::controller_path
   128:     def controller_path
(byebug) up

[18, 27] in /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/actionmailer-5.1.4/lib/action_mailer/rescuable.rb
   18:     end
   19:
   20:     private
   21:       def process(*)
   22:         handle_exceptions do
=> 23:           super
   24:         end
   25:       end
   26:   end
   27: end
(byebug) up

[10, 19] in /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/actionmailer-5.1.4/lib/action_mailer/rescuable.rb
   10:         rescue_with_handler(exception) || raise(exception)
   11:       end
   12:     end
   13:
   14:     def handle_exceptions #:nodoc:
=> 15:       yield
   16:     rescue => exception
   17:       rescue_with_handler(exception) || raise
   18:     end
   19:
(byebug) up

[17, 26] in /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/actionmailer-5.1.4/lib/action_mailer/rescuable.rb
   17:       rescue_with_handler(exception) || raise
   18:     end
   19:
   20:     private
   21:       def process(*)
=> 22:         handle_exceptions do
   23:           super
   24:         end
   25:       end
   26:   end
(byebug) up

[25, 34] in /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/actionview-5.1.4/lib/action_view/rendering.rb
   25:     include ActionView::ViewPaths
   26:
   27:     # Overwrite process to setup I18n proxy.
   28:     def process(*) #:nodoc:
   29:       old_config, I18n.config = I18n.config, I18nProxy.new(I18n.config, lookup_context)
=> 30:       super
   31:     ensure
   32:       I18n.config = old_config
   33:     end
   34:
(byebug) up

[604, 613] in /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/actionmailer-5.1.4/lib/action_mailer/base.rb
   604:         action: method_name,
   605:         args: args
   606:       }
   607:
   608:       ActiveSupport::Notifications.instrument("process.action_mailer", payload) do
=> 609:         super
   610:         @_message = NullMail.new unless @_mail_was_called
   611:       end
   612:     end
   613:
(byebug) up

[161, 170] in /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/activesupport-5.1.4/lib/active_support/notifications.rb
   161:         notifier.publish(name, *args)
   162:       end
   163:
   164:       def instrument(name, payload = {})
   165:         if notifier.listening?(name)
=> 166:           instrumenter.instrument(name, payload) { yield payload if block_given? }
   167:         else
   168:           yield payload if block_given?
   169:         end
   170:       end
(byebug) up

[16, 25] in /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/activesupport-5.1.4/lib/active_support/notifications/instrumenter.rb
   16:       # in the passed-in block.
   17:       def instrument(name, payload = {})
   18:         # some of the listeners might have state
   19:         listeners_state = start name, payload
   20:         begin
=> 21:           yield payload
   22:         rescue Exception => e
   23:           payload[:exception] = [e.class.name, e.message]
   24:           payload[:exception_object] = e
   25:           raise e
(byebug) up

[161, 170] in /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/activesupport-5.1.4/lib/active_support/notifications.rb
   161:         notifier.publish(name, *args)
   162:       end
   163:
   164:       def instrument(name, payload = {})
   165:         if notifier.listening?(name)
=> 166:           instrumenter.instrument(name, payload) { yield payload if block_given? }
   167:         else
   168:           yield payload if block_given?
   169:         end
   170:       end
(byebug) up

[603, 612] in /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/actionmailer-5.1.4/lib/action_mailer/base.rb
   603:         mailer: self.class.name,
   604:         action: method_name,
   605:         args: args
   606:       }
   607:
=> 608:       ActiveSupport::Notifications.instrument("process.action_mailer", payload) do
   609:         super
   610:         @_message = NullMail.new unless @_mail_was_called
   611:       end
   612:     end
(byebug) up

[100, 109] in /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/actionmailer-5.1.4/lib/action_mailer/message_delivery.rb
   100:     private
   101:       # Returns the processed Mailer instance. We keep this instance
   102:       # on hand so we can delegate exception handling to it.
   103:       def processed_mailer
   104:         @processed_mailer ||= @mailer_class.new.tap do |mailer|
=> 105:           mailer.process @action, *@args
   106:         end
   107:       end
   108:
   109:       def enqueue_delivery(delivery_method, options = {})
(byebug) up

[99, 108] in /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/actionmailer-5.1.4/lib/action_mailer/message_delivery.rb
    99:
   100:     private
   101:       # Returns the processed Mailer instance. We keep this instance
   102:       # on hand so we can delegate exception handling to it.
   103:       def processed_mailer
=> 104:         @processed_mailer ||= @mailer_class.new.tap do |mailer|
   105:           mailer.process @action, *@args
   106:         end
   107:       end
   108:
(byebug) up

[90, 99] in /Users/lokson/.rvm/gems/ruby-2.4.2@24pullrequests/gems/actionmailer-5.1.4/lib/action_mailer/message_delivery.rb
   90:     # Delivers an email:
   91:     #
   92:     #   Notifier.welcome(User.first).deliver_now
   93:     #
   94:     def deliver_now
=> 95:       processed_mailer.handle_exceptions do
   96:         message.deliver
   97:       end
   98:     end
   99:
(byebug) up

[255, 264] in /Users/lokson/24pullrequests/app/models/user.rb
   255:     return unless email_changed? && email.present?
   256:
   257:     generate_confirmation_token
   258:     self.confirmed_at = nil
   259:
=> 260:     ConfirmationMailer.confirmation(self).deliver_now
   261:   end
   262:
   263:   def generate_unsubscribe_token
   264:     self.unsubscribe_token ||= SecureRandom.uuid
(byebug) self
#<User id: nil, uid: 1, provider: "github", nickname: "eileen-beahan", email: "email1@factory.com", created_at: nil, updated_at: nil, gravatar_id: "dedrick_wisoky@wisozk.com", token: "97139babbec73b112ae5801ea2ce345d", email_frequency: nil, pull_requests_count: 0, last_sent_at: nil, twitter_token: nil, twitter_secret: nil, twitter_nickname: nil, confirmation_token: nil, confirmed_at: nil, name: nil, blog: nil, location: "Cambodia", thank_you_email_sent: false, lat: 0.40714353e2, lng: -0.74005973e2, ignored_organisations: [], unsubscribe_token: "ec388d37-7b77-4ee0-a308-fa6650b18797">
(byebug) generate_confirmation_token
nil
(byebug) self
#<User id: nil, uid: 1, provider: "github", nickname: "eileen-beahan", email: "email1@factory.com", created_at: nil, updated_at: nil, gravatar_id: "dedrick_wisoky@wisozk.com", token: "97139babbec73b112ae5801ea2ce345d", email_frequency: nil, pull_requests_count: 0, last_sent_at: nil, twitter_token: nil, twitter_secret: nil, twitter_nickname: nil, confirmation_token: nil, confirmed_at: nil, name: nil, blog: nil, location: "Cambodia", thank_you_email_sent: false, lat: 0.40714353e2, lng: -0.74005973e2, ignored_organisations: [], unsubscribe_token: "ec388d37-7b77-4ee0-a308-fa6650b18797">
(byebug) SecureRandom.urlsafe_base64
"AasHfiV6Wj6I87hOMFL4cA"
```

Reply:
Don't comment out failing tests. There was a change that did `generate_confirmation_token` as part of before_save/validation callback. Seems that it was moved to the email sending method. Test is checking for that method during `save`, so it's not being triggered. I wonder why it's still there.

(God, do I ever loathe RSpec with all the verbose trash it forces)

Reply:
After spending hours with this I commented this out to make it pending. How would you suggest to fix it?

Reply:
I'd remove it, and make sure that it's part of the `send_confirmation_email` test.

Reply:
Ha, it worked. Even though the context seems the same. Magic.